File: packages/dcf_primitives/ios/Classes/Components/DCFViewComponent.swift
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import UIKit
import dcflight

class DCFViewComponent: NSObject, DCFComponent {
    required override init() {
        super.init()
    }
    
    func createView(props: [String: Any]) -> UIView {
        let view = UIView()
        
        if let gradientData = props["backgroundGradient"] as? [String: Any] {
        }
        
        view.applyStyles(props: props)
        
        updateView(view, withProps: props)
        
        return view
    }
    
    func updateView(_ view: UIView, withProps props: [String: Any]) -> Bool {
        guard let view = view as? UIView else { return false }
        
        // Handle viewport detection (low-level API - any view can use it)
        // Uses DCFViewportObserver from dcflight framework
        let hasViewportCallbacks = props["onViewportEnter"] != nil || props["onViewportLeave"] != nil
        if hasViewportCallbacks {
            let viewportData = props["viewport"] as? [String: Any]
            let config = ViewportConfig(
                once: viewportData?["once"] as? Bool ?? false,
                amount: viewportData?["amount"] as? Double
            )
            DCFViewportObserver.shared.observe(view, config: config)
        } else {
            DCFViewportObserver.shared.unobserve(view)
        }
        
        if let gradientData = props["backgroundGradient"] as? [String: Any] {
            view.applyStyles(props: ["backgroundGradient": gradientData])
        }
        
        view.applyStyles(props: props)
        
        return true
    }
    
    func applyLayout(_ view: UIView, layout: YGNodeLayout) {
        view.frame = CGRect(x: layout.left, y: layout.top, width: layout.width, height: layout.height)
    }
    
    func viewRegisteredWithShadowTree(_ view: UIView, shadowView: DCFShadowView, nodeId: String) {
        objc_setAssociatedObject(view, 
                               UnsafeRawPointer(bitPattern: "nodeId".hashValue)!, 
                               nodeId, 
                               .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        
        // Setup viewport detection if callbacks are registered
        // Check props from associated object or wait for updateView
        DispatchQueue.main.async {
            // Viewport detection will be set up in updateView when props are available
        }
    }
    
    static func handleTunnelMethod(_ method: String, params: [String: Any]) -> Any? {
        return nil
    }
}

File: packages/dcf_primitives/android/src/main/kotlin/com/dotcorr/dcf_primitives/components/DCFViewComponent.kt
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.dotcorr.dcf_primitives.components

import android.content.Context
import android.graphics.PointF
import android.util.Log
import android.view.View
import com.dotcorr.dcflight.components.DCFComponent
import com.dotcorr.dcflight.components.DCFFrameLayout
import com.dotcorr.dcflight.components.DCFTags
import com.dotcorr.dcflight.extensions.applyStyles
import com.dotcorr.dcf_primitives.utils.DCFViewportObserver
import com.dotcorr.dcf_primitives.utils.ViewportConfig

class DCFViewComponent : DCFComponent() {

    companion object {
        private const val TAG = "DCFViewComponent"
    }

    override fun createView(context: Context, props: Map<String, Any?>): View {
        val view = DCFFrameLayout(context)

        view.setTag(DCFTags.COMPONENT_TYPE_KEY, "View")

        updateView(view, props)

        return view
    }

    override fun updateView(view: View, props: Map<String, Any?>): Boolean {
        // CRITICAL: Merge new props with existing stored props to preserve all properties
        val existingProps = getStoredProps(view)
        val mergedProps = mergeProps(existingProps, props)
        storeProps(view, mergedProps)
        
        // Handle viewport detection (low-level API - any view can use it)
        val hasViewportCallbacks = mergedProps.containsKey("onViewportEnter") || mergedProps.containsKey("onViewportLeave")
        if (hasViewportCallbacks) {
            val viewportData = mergedProps["viewport"] as? Map<*, *>
            val config = ViewportConfig(
                once = (viewportData?.get("once") as? Boolean) ?: false,
                amount = (viewportData?.get("amount") as? Number)?.toDouble()
            )
            DCFViewportObserver.observe(view, config)
        } else {
            DCFViewportObserver.unobserve(view)
        }
        
        val nonNullProps = mergedProps.filterValues { it != null }.mapValues { it.value!! }
        view.applyStyles(nonNullProps)
        return true
    }


    override fun getIntrinsicSize(view: View, props: Map<String, Any>): PointF {
        return PointF(0f, 0f)
    }

    override fun viewRegisteredWithShadowTree(view: View, nodeId: String) {
        Log.d(TAG, "View component registered with shadow tree: $nodeId")
        
        // Setup viewport detection if callbacks are registered
        val props = getStoredProps(view)
        val hasViewportCallbacks = props.containsKey("onViewportEnter") || props.containsKey("onViewportLeave")
        if (hasViewportCallbacks) {
            val viewportData = props["viewport"] as? Map<*, *>
            val config = ViewportConfig(
                once = (viewportData?.get("once") as? Boolean) ?: false,
                amount = (viewportData?.get("amount") as? Number)?.toDouble()
            )
            // Delay observation until view is laid out
            view.post {
                DCFViewportObserver.observe(view, config)
            }
        }
    }
    
    override fun handleTunnelMethod(method: String, arguments: Map<String, Any?>): Any? {
        return null
    }
}

